Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -2637,9 +2646,63 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     }
 }
 
+WKContentRuleListStore *WebExtensionContext::declarativeNetRequestRuleStore()
+{
+    String contentBlockerStorePath = m_extensionController->configuration().declarativeNetRequestStoreDirectory();
+    if (!FileSystem::makeAllDirectories(contentBlockerStorePath))
+        return nil;
+
+    return [WKContentRuleListStore storeWithURL:[NSURL fileURLWithPath:contentBlockerStorePath]];
+}
+
+void WebExtensionContext::removeDeclarativeNetRequestRules()
+{
+    if (!isLoaded())
+        return;
+
+    // Use all user content controllers in case the extension was briefly allowed in private browsing
+    // and content was injected into any of those content controllers.
+    auto allUserContentControllers = extensionController()->allUserContentControllers();
+
+    for (auto& userContentController : allUserContentControllers)
+        userContentController.removeContentRuleList(uniqueIdentifier());
+}
+
+void WebExtensionContext::addDeclarativeNetRequestRulesToPrivateUserContentControllers()
+{
+    [declarativeNetRequestRuleStore() lookUpContentRuleListForIdentifier:uniqueIdentifier() completionHandler:^(WKContentRuleList *ruleList, NSError *error) {
+        if (!ruleList)
+            return;
+
+        for (auto& controller : extensionController()->allPrivateUserContentControllers())
+            controller.addContentRuleList(*ruleList->_contentRuleList, m_baseURL);
+    }];
+}
+
 void WebExtensionContext::compileDeclarativeNetRequestRules(NSArray *rulesData)
 {
-    // FIXME: rdar://114823294 - Implement this.
+    // FIXME: rdar://118839289 - Avoid converting/compiling the rules data every time if the rules haven't changed.
+
+    NSArray<NSString *> *jsonDeserializationErrorStrings;
+    auto *allJSONObjects = [_WKWebExtensionDeclarativeNetRequestTranslator jsonObjectsFromData:rulesData errorStrings:&jsonDeserializationErrorStrings];
+
+    NSArray<NSString *> *parsingErrorStrings;
+    auto *allConvertedRules = [_WKWebExtensionDeclarativeNetRequestTranslator translateRules:allJSONObjects errorStrings:&parsingErrorStrings];
+
+    NSString *webKitRules = encodeJSONString(allConvertedRules, { JSONOptions::FragmentsAllowed });
```

Comment:
Can just be `JSONOptions::FragmentsAllowed` the `{ }` is not needed unless you have multiple options.

Reply:
Nit: `auto *`.